Add TabView SelectionPattern

chingucoding opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This PR adds the SelectionPattern to the TabViewAutomationPeer.
I am not using the internals ListView pattern since we plan to switch that out at some point, and then we would need to do all of that ourselves anyway.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Use the syntax "Closes #1234" or "Fixes #5678" so that GitHub will close the issue once the PR is complete. -->
Fixes #2847 
## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
Add API test
## Screenshots (if appropriate):
<!--- If you are making visual changes to a Control or adding a new Control, please include screenshots showing the result. -->

## Open questions
Given that the TabView now has the SelectionPattern, should the TabViewItems also have the SelectionItemPattern?

There is also a way to run Accessibility Insights as a test through the [Axe package](https://github.com/microsoft/axe-windows). Maybe we could leverage that to prevent those errors in the future? Given that the MUXControlsTestApp is a bit special compared to a regular app, maybe this could be done with the XCG, after all it contains a lot of samples for the controls.
@stmoy FYI

ranjeshj commented on dev/TabView/TabViewAutomationPeer.cpp:
```diff
@@ -18,6 +18,10 @@ TabViewAutomationPeer::TabViewAutomationPeer(winrt::TabView const& owner)
 // IAutomationPeerOverrides
 winrt::IInspectable TabViewAutomationPeer::GetPatternCore(winrt::PatternInterface const& patternInterface)
```
Yes, we will also need to update TabViewItemAutomationPeer to support SelectionItem. It currently derives from ListViewItem which will point to the ListView as the selection container which is not what we want here. I think we can do what we do in NavigationViewItem [here](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/NavigationView/NavigationViewItemAutomationPeer.cpp).

chingucoding commented:
Implemented SelectionItem pattern on the TabViewItem now, added new API test for reading the IsSelected property and changing selection through the peer. Tested manually that changing selection also get's correctly announced in narrator.

StephenLPeters commented:
/azp run

chingucoding commented:
Looks like last test run failed due to heap limit. Merged with master to get latest fix for that.

ranjeshj commented:
> Looks like last test run failed due to heap limit. Merged with master to get latest fix for that.

Could you please sync again? Keith made another fix.

chingucoding commented:
> Could you please sync again? Keith made another fix.

Done,

ranjeshj commented:
/azp run

chingucoding commented:
Seems that CI failed due to visual tree verification files: https://dev.azure.com/ms/microsoft-ui-xaml/_build/results?buildId=93164&view=logs&j=f1f05f7d-9051-5031-1c4e-31dcdce6d081&t=dc1e8999-6150-522b-4887-86903abea868

Is this because of the recent renaming of the folder? Or is it more of a "general" issue? Also why are there new AppBarButton files? @kaiguo FYI

StephenLPeters commented on dev/TabView/TabViewItem.cpp:
```diff
@@ -146,6 +151,16 @@ void TabViewItem::OnCloseButtonOverlayModeChanged(winrt::TabViewCloseButtonOverl
+
+void TabViewItem::SetParentTabView(winrt::TabView const& tabView)
+{
+    m_parentTabView = winrt::make_weak(tabView);
```
I think the fact that this is needed indicates that the components are architectured incorrectly.  The TabViewItem shouldn't need to know about the Parent TabView. Instead I think that it might have been better to thave selection status held on the TabViewItem, which can raise Selection changed events that the TabView can use to manage the single selection.  That being said I don't think we should change this now... maybe.

chingucoding replied:
That's a really good point. The question that comes to my mind now, is why the NavigationViewItem has an IsSelected property, while the TabViewItem does not. Both act more or less like a tab item, in both cases only a single item can be selected at a time. Was/is there a specific reasoning why those two diverge here?

ranjeshj replied:
TabViewItem does have an IsSelected property which is inherited from SelectorItem

chingucoding replied:
Oh missed that. The "Is selected" property doesn't appear in the official TabViewItem documentation. Also the TabViewItem/TabView do not care about that property at all.

Was this by design? I think @StephenLPeters is right here, that is something we should update. Currently only TabView knows about selection in contrast to NavigationView, where the items also know their selection status.

Felix-Dev replied:
> Oh missed that. The "Is selected" property doesn't appear in the official TabViewItem documentation. Also the TabViewItem/TabView do not care about that property at all.

The TabViewItem.IsSelected property does work though. See this XAML:
```xml
<muxc:TabView>
    <muxc:TabView.TabItems>
        <muxc:TabViewItem Header="Tab 1"/>
        <muxc:TabViewItem Header="Tab 2"/>
        <muxc:TabViewItem Header="Tab 3" IsSelected="True"/>
        <muxc:TabViewItem Header="Tab 4"/>
    </muxc:TabView.TabItems>
</muxc:TabView>
```
This gives me the expected result:
![image](https://user-images.githubusercontent.com/1398851/87227520-e0f94080-c39b-11ea-907a-24fd876b6237.png)

chingucoding replied:
So is ListView doing selection for us here then?

Felix-Dev replied:
I would assume so, yes. Same as with TreeView when we are in TreeView.SelectionMode = Single.

Felix-Dev replied:
The incomplete WinUI documentation is an issue though. @ranjeshj @StephenLPeters Can the WinUI API documentation be updated to also include all the inherited APIs like the OS XAML documentation? I aleady had to open the OS XAML documentation quite a few times for WinUI components (like the NavigationView) because that documentation also lists all the inherited APIs.

chingucoding replied:
Since IsSelected works here, I removed the "GetParentTabView" pattern/code.

StephenLPeters replied:
@chigy and @YuliKl for the documentation question

StephenLPeters replied:
@chingucoding would it makes sense to move the pointer to the parent tab view onto the TabViewItemAutomationPeer since the TabViewItem itself doesn't need it?

chingucoding replied:
But how would we set that property then? After all, we don't know when the AutomationPeer for a TabViewItem will be created, so TVI needs to "save" it until it will be needed. And since AutomationPeer are more or less "fire and forget" anyway (meaning they get created and can be used only once and get recreated for the next usage), the TVI AutomationPeer would always have to ask a TVI for it's parent TabView.

StephenLPeters replied:
makes sense

ranjeshj commented on dev/TabView/TabViewItemAutomationPeer.cpp:
```diff
@@ -43,3 +51,37 @@ winrt::hstring TabViewItemAutomationPeer::GetNameCore()
+
+winrt::IRawElementProviderSimple TabViewItemAutomationPeer::SelectionContainer()
+{
+    return nullptr;
```
You will need to return the TabView's automation peer here.

chingucoding replied:
So we need a reference to TabView here anyway?

ranjeshj replied:
Good point. We need to either pass it down to the item or look up the tree from the item in this case. That said, I'm not exactly sure how Narrator uses this method though.

chingucoding replied:
I've added the logic back, I added earlier in this PR. Walking up the tree seems a bit expensive here, though I am not a fan of having this relation between TabViewItem and TabView just for this.

StephenLPeters replied:
@YuliKl is there someone we can ask about how Narrator uses this?

LukaszMakar replied:
Narrator uses `SelectionContainer` for a few purposes:
1. Listening to UIA events for the entire container.
2. Understanding how selection behaves in the container and therefore which actions should be presented to the user. For instance, if the container is `single select + required`, Narrator will not expose the action of removing an item from selection.
3. Making announcements that inform the user about selection state of individual items. For instance, Narrator's "selected" for lists depends on whether the container supports multi-selection. Not quite your case but it does not mean it cannot become one.

Does that help?

chingucoding replied:
I think point 2 is a good reason here to not provide a `nullptr` for the `SelectionContainer`. After all TabView does not allow deselecting an item, and without the `SelectionContainer` being provided, Narrator wouldn't be able to hide that option from users.

Thank you for the information @LukaszMakar !

StephenLPeters commented:
/azp run

StephenLPeters commented:
/azp run
